thali/thalireplicationmanager.js

```diff
@@ -0,0 +1,301 @@
+'use strict';
+
+var ThaliEmitter = require('./thaliemitter');
+var EventEmitter = require('events').EventEmitter;
+var inherits = require('util').inherits;
+var net = require('net');
+var multiplex = require('multiplex');
+var validations = require('./validations');
+
+var e = new EventEmitter();
+
+var PEER_AVAILABILITY_CHANGED = ThaliEmitter.events.PEER_AVAILABILITY_CHANGED;
+var NETWORK_CHANGED = ThaliEmitter.events.NETWORK_CHANGED;
+
+inherits(ThaliReplicationManager, EventEmitter);
+
+function ThaliReplicationManager(db, emitter) {
+  this._db = db;
+  this._emitter = (emitter || new ThaliEmitter());
+  this._replications = {};
+  this._clients = {};
+  this._isStarted = false;
+  this._serverBridge = null;
+  this._serverBridgePort = 0;
+  EventEmitter.call(this);
+}
+
+ThaliReplicationManager.events = {
+  STARTING: 'starting',
+  STARTED: 'started',
+  STOPPING: 'stopping',
+  STOPPED: 'stopped',
+  START_ERROR: 'startError',
+  STOP_ERROR: 'stopError',
+  CONNECT_ERROR: 'connectError',
+  DISCONNECT_ERROR: 'disconnectError',
+  SYNC_ERROR: 'syncRetry'
+};
+
+/**
+* Starts the Thali replication manager with the given device name and port number
+* @param {String} deviceName the device name to advertise.
+* @param {Number} port the port number used for synchronization.
+*/
+ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
+  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+  validations.ensureValidPort(port);
+  validations.ensureNonNullOrEmptyString(dbName, 'dbName');
+
+  this.emit(ThaliReplicationManager.events.STARTING);
+
+  this._port = port;
+  this._deviceName = deviceName;
+  this._dbName = dbName;
+  this._serverBridge = muxServerBridge.call(this, port);
+  this._serverBridge.listen(function () {
+    this._serverBridgePort = this._serverBridge.address().port;
+
+    this._emitter.startBroadcasting(deviceName, this._serverBridgePort, function (err) {
+      if (err) {
+        this._isStarted = false;
+        this.emit(ThaliReplicationManager.events.START_ERROR, err);
+      } else {
+        this._isStarted = true;
+        this._emitter.addListener(PEER_AVAILABILITY_CHANGED, syncPeers.bind(this));
+        this._emitter.addListener(NETWORK_CHANGED, networkChanged.bind(this));
+        this.emit(ThaliReplicationManager.events.STARTED);
+      }
+    }.bind(this));
+  }.bind(this));
+};
+
+/**
+* Starts the Thali replication manager with the given device name and port number
+* @param {String} deviceName the device name to advertise.
+* @param {Number} port the port number used for synchronization.
+*/
+ThaliReplicationManager.prototype._restart = function () {
```

Comment:
Who calls this? I did a search for _restart and only found the method declaration. Also if someone does call restart how can we be sure that the existing serverBridge has been closed? Shouldn't we try to close it again (with a try/catch) if it's not null just to be sure we aren't leaking anything?